tagless-final/src/main/java/com/github/jorgecastillo/kotlinandroid/tagless/runtime/instances/Instances.kt

```diff
@@ -1,27 +1,17 @@
 package com.github.jorgecastillo.kotlinandroid.tagless.runtime.instances
 
-import arrow.effects.IO
-import arrow.effects.IOHK
-import arrow.instance
-import com.github.jorgecastillo.kotlinandroid.tagless.algebras.persistence.DataSource
-import com.github.jorgecastillo.kotlinandroid.tagless.algebras.persistence.HeroesRepository
-import com.github.jorgecastillo.kotlinandroid.tagless.algebras.business.HeroesUseCases
-import com.github.jorgecastillo.kotlinandroid.tagless.algebras.ui.Navigation
+import arrow.dagger.effects.instances.ArrowEffectsInstances
+import arrow.dagger.instances.ArrowInstances
+import arrow.effects.ForIO
 import com.github.jorgecastillo.kotlinandroid.tagless.algebras.ui.Presentation
+import dagger.Component
 
-typealias TargetRuntime = IOHK
+@Component(modules = [
+    ArrowInstances::class,
+    ArrowEffectsInstances::class
+])
+interface Instances {
+    fun presentation(): Presentation<ForIO>
+}
 
-@instance(IO::class)
-interface IOPresentationInstance<F> : Presentation<F>
-
-@instance(IO::class)
-interface IONavigationInstance<F> : Navigation<F>
-
-@instance(IO::class)
-interface IOHeroesUseCasesInstance<F> : HeroesUseCases<F>
-
-@instance(IO::class)
-interface IOHeroesRepositoryInstance<F> : HeroesRepository<F>
-
-@instance(IO::class)
-interface IODataSourceInstance<F> : DataSource<F>
+fun instances(): Instances = DaggerInstances.create()
```

Comment:
How many instances is this creating? I guess all the ones defined on the component attached modules for the IO type?

Is this method **mandatory** to do the linking so dagger becomes able to generate the instances at compile time? I guess it is, since it is the only way Dagger has to know which is the return type expected here and therefore crawl using the compile time reflection API to get the `@component` annotation from the type.

So, I see another de-linking readability problem here, since this method feels needed to me by thinking a bit, but the user does not even know when reading the code. Feels like he could directly call `DaggerInstances.create()` from the Activities, the same way we do with the very usual dagger `component.inject(this)` (that's what the end user is used to) and remove the redundant `instances()` method. But that would break the build since dagger would not have a way to know which @component corresponds to this class (since there is no `Instances` type anymore around) ?

Also, where are the the instances for `Navigation<F>`, `HeroesUseCases<F>`, `HeroesRepository<F>`, and `DataSource<F>`?

Reply:
This can be just a `val`. 
This is not like `@Provides` since this is the point where the entire graph is materialized and made concrete. Note Arrow has not changed anything to bring Dagger Support and this is just regular Dagger 2 usage. The explicit declaration of `Navigation<F>` etc... are no longer necessary because Dagger 2 auto instantiates and provides instances for all classes whose constructor is annotated wit `@Inject`.

Alternatively a user may declare:
```
@typeclass Service<F> : TC { 
  ... 
}
class ServiceInstance<F> @Inject constructor(val deps...): Service<F> { 
... 
}

@Module abstract class LocalInstances<F> {
  @Provides
  fun service(ev: ServiceInstance<F>): Service<F> = ev
}

@Module class IOLocalInstances: LocalInstances<ForIO>

@Component(modules = [
  ArrowInstances:class,
  ArrowEffectsInstances::class
  IOLocalInstances::class
])
interface Edge
```
But all the `@Provides` is no necessary unless your type classes provide abstract members that require someone to define how the instances is create.

Dagger 2 does not require `component.inject(this)` or any of those idioms if you can prove all your instances are in scope in the imported modules. You can materialize them with abstract methods in the @Component.

Also the usage of Dagger in Arrow is not yet documented because it had not been finished. I have not introduced syntax in your project yet because I wanted to just get it to a compilable state but syntax is next and that requires the type classes to be defined as interfaces if we find that any of the method defined in your services are subject to be exposed as syntax over `Kind<F, ?>`